### DISCUSSION TITLE
Inheriting from open definition causes a lack of concreteness

### DISCUSSION THREAD

**Discussion by NotTheEconomist** [cue-lang/cue#1658] [Q&A]
I'm a bit confused, and wonder if I'm running up against a bug in the language. My actual example is writing a [semver][1] schema where I can define a semver object either by a structured object

```
#Semver & {
  version: {
    major: 1
    minor: 0
    patch: 0
    prerelease: ""
    metadata: ""
  }
}
```

but also with a string and parse that object from the regex

```
#Semver & {
  string: "1.0.0"
}
```

To capture both those use cases I have a parent of `#Semver` which is a definition that looks like:

```
#VersionBase: {
  _version: _  // more than just semver will inherit from this
  _string: string
  ...
}
```

but when defining `#Semver` as `#Semver: #VersionBase & { // etc`, things that were once concrete end up as unresolved disjunctions.

Here's a simplified example on Play https://cuelang.org/play/?id=4rQRf7UoAzl#cue@export@cue

```
import (
    "strconv"
)

#Base: {
    x_a: string
    x_i: int
    ...
}

#MyType: {
    {
        a: string
        x_a: a
        x_i: strconv.Atoi(a)
    } | {
        i: int
        x_a: strconv.FormatInt(i, 10)
        x_i: i
    }
}

#MyIdenticalType: #Base & {
    {
        a: string
        x_a: a
        x_i: strconv.Atoi(a)
    } | {
        i: int
        x_a: strconv.FormatInt(i, 10)
        x_i: i
    }
}

m: #MyType & { a: "1" }
o: #MyType & { i: 1 }

m2: #MyIdenticalType & { a: "1" }
o2: #MyIdenticalType & { i: 1 }
```

`m` and `o` appear to be concrete, while `m2` and `o2` keep their disjunction.

[1]:https://semver.org/

**Comment by eonpatapon** [comment]
In the case of `m2` or `o2`, because `#Base` makes the definition open when you specify any attribute `cue` cannot know if you want to specialize some part of the disjunction or if you want to just add another field unrelated to the disjunction. 

The result of `m2` shows:

```cue
m2: {
	a:   "1"
	x_a: "1"
	x_i: 1
} | {
	a:   "1"
	i:   int
	x_a: string & strconv.FormatInt(i, 10)
	x_i: int
}
```

We can see that `a` has been added in all elements of the disjunction because `#MyIdenticalType` is open. For `cue`, all elements of the disjunction are valid values even if some is not concrete. For `cue` to eliminate a value from the disjunction there must be an error, which isn't the case here, therefore all values are kept. 

To disambiguate the disjunction we would need to specify a field which would produce an error in either side of the disjuction, for example:

```cue
m2: #MyIdenticalType & { a: "1", i: null }

m2: {
	a:   "1"
	x_a: "1"
	x_i: 1
	i:   null
}
```

This is probably not what you want in the end.

A possibility here would be to make `#Base` closed and embed it in `#MyIdenticalType`. This way `#MyIdenticalType` will stay closed: https://cuelang.org/play/?id=GeJ06T_j-f-#cue@export@cue

**Reply by myitcv** [reply]
To build on @eonpatapon's answer, another way of achieving the reuse you want here is to [embed](https://tip.cuelang.org/docs/references/spec/#embedding) `#Base`:

https://cuelang.org/play/?id=x-oTcDkUlBT#cue@export@cue

Per the spec, the closedness constraints on `#Base` are relaxed for the purposes for embedding, allowing the disjunction you have in `#MyIdenticalType`, but then the result is closed.
